src/committee/orchestrator.rs

```diff
@@ -65,191 +292,233 @@ impl Orchestrator {
-        for (member_id, member) in &threshold_of_members {
-            // send json RPC request
-            let rpc_ctx = RpcCtx::new(Some("2.0"), None, Some(member.address.clone()), None, None);
-            let resp = json_rpc_request(
-                &rpc_ctx,
-                "round_2_signing",
-                &[serde_json::value::to_raw_value(&round2_request)?],
-            )
-            .await;
-            debug!("resp to 2nd request: {:?}", resp);
-
-            let resp = resp.context("second rpc request to committee didn't work")?;
-
-            let response: bitcoincore_rpc::jsonrpc::Response = serde_json::from_str(&resp)?;
-            let round2_response: Round2Response = response.result()?;
+        'retry: loop {
+            //
+            // Round 1
+            //
+
+            let mut commitments_map = BTreeMap::new();
+
+            let mut available_members = {
+                let ms_r = member_status.read().unwrap();
+                self.committee_cfg
+                    .members
+                    .iter()
+                    .filter(|(key, _)| ms_r.get_member_status(key) == MemberStatus::Online)
+                    .collect_vec()
+            };
+            if available_members.len() < self.committee_cfg.threshold {
+                return Err(anyhow::Error::msg("not enough available signers"));
+            }
 
-            // store the commitment
-            signature_shares.insert(**member_id, round2_response.signature_share);
-        }
+            available_members.shuffle(&mut rand::thread_rng());
+            available_members.truncate(self.committee_cfg.threshold);
+
+            let futures = available_members
+                .iter()
+                .map(|(_, member)| async {
+                    let rpc_ctx =
+                        RpcCtx::new(Some("2.0"), None, Some(member.address.clone()), None, None);
+                    json_rpc_request(
+                        &rpc_ctx,
+                        "round_1_signing",
+                        &[serde_json::value::to_raw_value(&bob_request).unwrap()],
+                    )
+                    .await
+                })
+                .collect_vec();
+
+            let round_1_responses = join_all(futures).await;
+
+            for (idx, resp) in round_1_responses.into_iter().enumerate() {
+                let (member_id, member) = available_members[idx];
+                debug!("resp to 1st request from {:?}: {:?}", member_id, resp);
+                let resp = match resp {
+                    Ok(x) => x,
+                    Err(rpc_error) => {
+                        warn!("Round 1 error with {}, marking as disconnected and retrying round 1: {rpc_error}", member.address);
+                        let mut ms_w = member_status.write().unwrap();
+                        ms_w.mark_as_disconnected(member_id);
+                        continue 'retry;
```

Comment:
You're right! We can definitely do that. Right now it's fine, because even with like 5-10 MPC nodes will be fast enough, but when there will be more nodes the requests that already succeeded are redundant.